[TSM] Panic during compaction

Doing a full-write load tests across two different servers, I encountered this panic:
```
panic: runtime error: slice bounds out of range

goroutine 7044023 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*indirectIndex).UnmarshalBinary(0xc84de30320, 0x7f7925d7b1fa, 0x166b6, 0x166be, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:622 +0x532
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*mmapAccessor).init(0xc84fe81d60, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:858 +0x57c
github.com/influxdb/influxdb/tsdb/engine/tsm1.NewTSMReaderWithOptions(0x0, 0x0, 0xc8b7d9f040, 0x0, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/reader.go:148 +0x2d8
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*FileStore).Replace(0xc82f7440e0, 0xc829f0dd00, 0x2, 0x2, 0xc8c0cc9a10, 0x1, 0x1, 0x0, 0x0)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/file_store.go:367 +0x41f
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).compactTSMLevel.func1(0xc89597bfd0, 0xc820cd6f30, 0x1, 0x200010001, 0x0, 0xc829f0dd00, 0x2, 0x2)
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:524 +0x79d
created by github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).compactTSMLevel
        /tmp/tmp.m6jFTgA2aM/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:535 +0x25f
```

Interestingly, it occurred on both boxes at nearly the same time - 4:47am on one and 4:52 on the other, which was ~16h into the test. Both tests were started at the same time. The first was 100k series using 5k batches, and 1m spread between timestamps (37GB written, 164 TSM files). The other was the same, but with 50k series (36GB written, 105 TSM files).

This test was performed with the following nightly build:
```
InfluxDB v0.10.0-nightly-6ccc416 (git: master 6ccc416ef0d3b0ba6d43328a2719886b66eb5e49, built 2016-01-01T05:00:47+0000)
```

--- a/tsdb/engine/tsm1/reader.go
+++ b/tsdb/engine/tsm1/reader.go
@@ -607,7 +607,7 @@ func (d *indirectIndex) UnmarshalBinary(b []byte) error {
 		i++
 
 		// 2 byte count of index entries
-		count := btou16(b[i : i+indexCountSize])
+		count := int32(btou16(b[i : i+indexCountSize]))
 		i += indexCountSize
 
 		// Find the min time for the block

--- a/tsdb/engine/tsm1/reader_test.go
+++ b/tsdb/engine/tsm1/reader_test.go
@@ -739,6 +739,44 @@ func TestBlockIterator_Sorted(t *testing.T) {
 	}
 }
 
+func TestIndirectIndex_UnmarshalBinary_BlockCountOverflow(t *testing.T) {
+	dir := MustTempDir()
+	defer os.RemoveAll(dir)
+	f := MustTempFile(dir)
+	defer f.Close()
+
+	w, err := tsm1.NewTSMWriter(f)
+	if err != nil {
+		t.Fatalf("unexpected error creating writer: %v", err)
+	}
+
+	for i := 0; i < 3280; i++ {
+		w.Write("cpu", []tsm1.Value{tsm1.NewValue(time.Unix(int64(i), 0), float64(i))})
+	}
+
+	if err := w.WriteIndex(); err != nil {
+		t.Fatalf("unexpected error closing: %v", err)
+	}
+
+	if err := w.Close(); err != nil {
+		t.Fatalf("unexpected error closing: %v", err)
+	}
+
+	f, err = os.Open(f.Name())
+	if err != nil {
+		t.Fatalf("unexpected error open file: %v", err)
+	}
+
+	r, err := tsm1.NewTSMReaderWithOptions(
+		tsm1.TSMReaderOptions{
+			MMAPFile: f,
+		})
+	if err != nil {
+		t.Fatalf("unexpected error created reader: %v", err)
+	}
+	defer r.Close()
+}
+
 func BenchmarkIndirectIndex_UnmarshalBinary(b *testing.B) {
 	index := tsm1.NewDirectIndex()
 	for i := 0; i < 100000; i++ {

